### DISCUSSION TITLE
Breakage.yml

### DISCUSSION THREAD

**Discussion by ocots** [JuliaSmoothOptimizers/Organization#53] [General]
Hi!

@amontoison The [Breakage.yml](https://github.com/JuliaSmoothOptimizers/ADNLPModels.jl/blob/main/.github/workflows/Breakage.yml) gives info about failure or success for dependent packages / applications.

Based on your file, I have made [one](https://github.com/control-toolbox/OptimalControl.jl/blob/main/.github/workflows/BreakageApplication.yml) where applications are tested, in the sense that it is not the unit tests which are run but the documentation. Here is an [example](https://control-toolbox.org/calculus_of_variations) of such an application. The dependence to our package OptimalControl.jl may be seen in the [`docs/Project.toml`](https://github.com/control-toolbox/calculus_of_variations/blob/main/docs/Project.toml).

Here is my question. 

Would it be possible to add the following info in the comment from the bot: the version of our package that is actually used by the package / application which is tested?

Thanks.

**Comment by amontoison** [comment]
Hi @ocots!

Yes, you should be able to do that. You just need to use the badge of the package in the `breakage.yml`.
The modification should here:
https://github.com/JuliaSmoothOptimizers/ADNLPModels.jl/blob/main/.github/workflows/Breakage.yml#L120

The URL of the badge with the last release of a CUTEst.jl is:
```shell
https://img.shields.io/github/release/JuliaSmoothOptimizers/CUTEst.jl.svg
```
You can adapt it dor any package in the organization `control-toolbox`.
Example:
```shell
https://img.shields.io/github/release/control-toolbox/OptimalControl.jl.svg
```

**Comment by ocots** [comment]
My question was not clear.

I want to get the version of OptimalControl.jl used by the application that I test which should be written in the Project.toml. 

For instance, [here](https://github.com/control-toolbox/calculus_of_variations/blob/main/docs/Project.toml) I want to get `0.11`. 

But I think I have answered to my question since I can parse the file and then add some info here

https://github.com/control-toolbox/OptimalControl.jl/blob/676737ab8772ecbaaaac09c1b251ece6f7f0b49e/.github/workflows/BreakageApplication.yml#L81

I try it.

**Reply by amontoison** [reply]
@ocots 
You will only have a lower bound if parse your `Project.toml` and not the current version of your packages.
I suggest that:
```shell
version=$(julia --project -e 'using Pkg, Logging; with_logger(NullLogger()) do; Pkg.installed()["OptimalControl"] |> println; end')
echo $version
```
It's an example with `OptimalControl.jl` but you can adapt it for any package.

**Reply by ocots** [reply]
Note that the `Project.toml` is in the `docs` directory for the application I test breakage. Actually, for the application we use Documenter.jl to generate some documentation. With my breakage, I test if the documentation is generating well or if there is a failure.

Should I update with:

```bash
version=$(julia --project=docs/ -e 'using Pkg, Logging; with_logger(NullLogger()) do; Pkg.installed()["OptimalControl"] |> println; end')
echo $version
```

**Reply by ocots** [reply]
This seems to work:

```bash
COMPAT_VERSION=$(julia --project=docs/ -e 'using Pkg, Logging; with_logger(NullLogger()) do; Pkg.instantiate(); Pkg.installed()["OptimalControl"] |> println; end')
```

**Reply by amontoison** [reply]
Send me the link of the comment of the bot if it works. I'm curious to see the result.

**Reply by ocots** [reply]
I just have a last wish. OptimalControl is the package from where I am doing breakage. Can I get the name somehow?

**Reply by ocots** [reply]
From this line 

https://github.com/JuliaSmoothOptimizers/ADNLPModels.jl/blob/75a15faeb97bfcadcb5dccadbf17f62de35f21a9/.github/workflows/Breakage.yml#L81

it seems that the name of the package should be in the path. Better idea?

**Reply by amontoison** [reply]
I think that `package=$(ls ../..)` is enough.

**Reply by ocots** [reply]
To get the name, I have written:

```bash
PKG_SRC_PATH=`pwd`
PKG_SRC_NAME=`basename -s ".jl" $PKG_SRC_PATH`
```

**Comment by ocots** [comment]
@amontoison You can find my [BreakageApplications.yml](https://github.com/control-toolbox/OptimalControl.jl/blob/main/.github/workflows/BreakageApplication.yml) and a result [here](https://github.com/control-toolbox/OptimalControl.jl/pull/334#issuecomment-2313566236).

It seems to work. Note that all applications fail since I have updated the version. The current version of OptimalControl is 0.12.

**Comment by ocots** [comment]
I don't want to run this breakage every time I make a PR or push on main so I think next time I will add a 

```yml
on:
    workflow_dispatch:
```

to run it whenever I want and instead of creating a comment in a PR, I would rather it creates an issue with the output.

**Reply by amontoison** [reply]
I suggest to trigger this breakage with a label.
It's a neat solution.
It's what we use in `ADNLPModels.jl` when we want to run a specific benchmark (Hessian, Jacobian, gradient, etc...):
https://github.com/JuliaSmoothOptimizers/ADNLPModels.jl/blob/main/.github/workflows/BenchmarkJacobian.yml#L11

Based on the results, you can create an issue with this GitHub Action: https://github.com/JasonEtco/create-an-issue

**Reply by ocots** [reply]
Very nice solution. It is better! 

It works well: https://github.com/control-toolbox/OptimalControl.jl/pull/336.

**Reply by ocots** [reply]
No need of issue for the moment.

**Comment by tmigot** [comment]
@ocots Can we close this?

**Reply by ocots** [reply]
@tmigot Yes we can. 

An example where I show compat when I test on packages and not applications :
 https://github.com/control-toolbox/CTBase.jl/blob/main/.github/workflows/Breakage.yml
